msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
+                "Please use Demography.sample() to generate a list of samples "
+                "for your model, which can be used instead."
+            )
+        _insert_integer_samples(n, ploidy, tables)
+    else:
+        raise TypeError(
+            "The samples argument must be either an integer (for single population "
+            "models) or a list of msprime.Sample objects. The Demography.sample() "
+            "function is useful for generating samples for complex demographic "
+            "models."
+        )
+
+
+def _parse_sim_ancestry(
+    samples=None,
     *,
-    ploidy=None,
     sequence_length=None,
-    samples=None,
-    discrete_genome=None,
     recombination_rate=None,
+    gene_conversion_rate=None,
+    gene_conversion_track_length=None,
+    discrete_genome=None,
     population_size=None,
     demography=None,
+    ploidy=None,
     model=None,
+    initial_state=None,
+    start_time=None,
+    end_time=None,
+    record_migrations=None,
+    record_full_arg=None,
+    num_labels=None,
     random_seed=None,
+    random_generator=None,
     num_replicates=None,
+    replicate_index=None,
 ):
-    if discrete_genome is None:
-        discrete_genome = True
-    if ploidy is None:
-        ploidy = 1
-    if population_size is None:
-        population_size = 1
-    # Samples are interpreted as k-ploid individuals here. For the
-    # initial hack we just multiply by this value.
-    if sample_size is not None:
-        sample_size *= ploidy
-    elif samples is not None:
-        old_samples = samples
-        samples = [None] * (len(samples) * ploidy)
-        j = 0
-        for sample in old_samples:
-            for _ in range(ploidy):
-                samples[j] = sample
-                j += 1
-
-    recomb_map = None
-    if recombination_rate is None:
-        if sequence_length is None:
-            # We've no recombination, so this is a single locus simulation.
-            # Defaulting to length of 1 is fine.
+    """
+    Argument parser for the sim_ancestry frontend. Interprets all the parameters
+    and returns an appropriate instance of Simulator.
+    """
+    # As a general rule we try to cast any input value to the required types
+    # early and in a way that provides an interpretable traceback.
+
+    # Simple defaults.
+    start_time = 0 if start_time is None else float(start_time)
+    end_time = sys.float_info.max if end_time is None else float(end_time)
+    # Note: casting to bool here means that we support Truthiness, *except* for
+    # None, which is interpreted as "give me the default". This may lead to
+    # confusing problems for (misguided) users of truthiness.
+    discrete_genome = True if discrete_genome is None else bool(discrete_genome)
+    record_migrations = False if record_migrations is None else bool(record_migrations)
+    record_full_arg = False if record_full_arg is None else bool(record_full_arg)
```

Comment:
Could use a coercion function like this?
```python
def coerce_tristate(b, meaning_of_None):
    if b is None:
        return meaning_of_None
    if not isinstance(b, bool):
        raise TypeError("expected True, False, or None")
    return b

# truth table
assert(True == coerce_tristate(None, True))
assert(True == coerce_tristate(True, True))
assert(False == coerce_tristate(False, True))
assert(False == coerce_tristate(None, False))
assert(True == coerce_tristate(True, False))
assert(False == coerce_tristate(False, False))
```

Reply:
As in, enforce lack of truthiness, so people have to cast to bool explicitly themselves if they want it? Good idea...

Reply:
Done, ta.